### PR TITLE
workflows: staging build and test

### PR DESCRIPTION
**Pull request by patrick-stephens**
Addresses #3753 

New workflows added to automate the build and test of releases using the new `staging` environment.
No changes made to current process to ensure we can keep using it.

Build & test of packaging
---
Packages built to staging in S3 bucket: https://fluentbit-staging.s3.amazonaws.com
We then verify the packages using kitchen-dokken to spin up OS images as containers, install the relevant RPM/Deb and check the service is properly running then. We are testing that the packaging process is correct.

Containers build to Github Container Registry, gchr.io, using multi-arch manifests.
Container tests then verify each architecture runs locally as well as a simple Helm deployment on KIND.

All package and container build definitions brought into the repo from external sources - containers were in this repo and packages were not so that is now identical plus having them together makes it a lot easier to manage and use.

Security
---
Trivy and Dockle scanning added - ignores current failures so these should be reviewed and addressed as needed. 
Hadolint and Shellcheck really should be used too but this can be a separate PR.

Cosigning of container images if a key is provided, and using the experimental keyless option too.
GPG signing of binary packages as well as normal.

Additional work
---
Initial promotion from staging to release provided using a new `release` environment for approval - this needs creating.
Initial multi-arch container image definition and workflow also added.

Follow up PRs to improve testing, build on self-hosted and cover the promotion to release process. Trying to prevent a big bag and reduce review overhead.

Infra updates
----
Create `release` and `staging` environments.
Create the following secrets:
* AWS_S3_BUCKET_STAGING
* AWS_S3_BUCKET_RELEASE
* AWS_ACCESS_KEY_ID
* AWS_SECRET_ACCESS_KEY
* COSIGN_PRIVATE_KEY
* COSIGN_PASSWORD - optional if private key does not require
* COSIGN_PUBLIC_KEY
* FLUENTBITIO_HOST
* FLUENTBITIO_USERNAME
* FLUENTBITIO_SSHKEY
* GPG_PRIVATE_KEY

We can actually start breaking these secrets up into the two environments.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [NA] Example configuration file for the change
- [NA] Debug log output from testing the change

- [NA] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [NA] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

File: tools/act/run.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    -s DOCKERHUB_ORGANIZATION="$DOCKERHUB_ORGANIZATION" \
+    -s AWS_ACCESS_KEY_ID="$AWS_ACCESS_KEY_ID" \
+    -s AWS_SECRET_ACCESS_KEY="$AWS_SECRET_ACCESS_KEY" \
+    -s AWS_S3_BUCKET_STAGING="$AWS_S3_BUCKET_STAGING" \
+    -s AWS_S3_BUCKET_RELEASE="$AWS_S3_BUCKET_RELEASE" \
+    -s GPG_PRIVATE_KEY="$GPG_PRIVATE_KEY" \
+    -s FLUENTBITIO_HOST="$FLUENTBITIO_HOST" \
+    -s FLUENTBITIO_USERNAME="$FLUENTBITIO_USERNAME" \
+    -s FLUENTBITIO_SSHKEY="$FLUENTBITIO_SSHKEY" \
+    --env AWS_S3_ENDPOINT="http://localhost:9000" \
```
Comment (gagoar) [inline tools/act/run.sh:34]: can this one be a env variable as well?

File: packaging/update-repos.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+for DEB_REPO in "${DEB_REPO_PATHS[@]}"; do
+    REPO_DIR=$(realpath -sm "$BASE_PATH/$DEB_REPO" )
+    [[ ! -d "$REPO_DIR" ]] && continue
+
+    echo "Updating $DEB_REPO"
+
+    find "$REPO_DIR" -name "td-agent-bit-${VERSION}_*.deb" -exec debsigs --sign=origin -k "$GPG_KEY" {} \;
+    dpkg-scanpackages -m "$REPO_DIR" | gzip -c > "$REPO_DIR"/Packages.gz
+done
```
Comment (gagoar) [inline packaging/update-repos.sh:57]:
```suggestion
done

```

Comment (patrick-stephens) [reply packaging/update-repos.sh:57]: Ah linting should have got that! :+1:

File: tools/act/run.sh (hunk 1 above)
Comment (patrick-stephens) [reply tools/act/run.sh:34]: yeah probably, to be honest I might just review the script as Act won't work in a lot of ways now!

Comment (gagoar) [reply tools/act/run.sh:34]: make sense, maybe a note around what can be use to test the workflows will suffice ?

File: packaging/update-repos.sh (hunk 2 above)
Comment (gagoar) [reply packaging/update-repos.sh]: yea, I saw it in other files. It seems that it has happened in every file you have created from scratch. I use a setting to make my editor just add the new line in every file.

**Review by niedbalski — CHANGES_REQUESTED** [review]
* git rebase HEAD~250 --signoff
* update the workflows/README.md
* provide a list of the required secrets to be added into infra.
* remove act (?)

**Comment by patrick-stephens** [comment]
Closed as original repo had issues with squash+sign

**Pull request closed without merging**